src/aiohubspace/v1/auth.py

```diff
@@ -1,342 +1,361 @@
-    try:
-        login_url: str = login_form.attrs["action"]
-    except KeyError:
-        raise InvalidResponse("Unable to extract login url")
-    parsed_url = urlparse(login_url)
-    login_data = parse_qs(parsed_url.query)
-    try:
-        return auth_sess_data(
-            login_data["session_code"][0],
-            login_data["execution"][0],
-            login_data["tab_id"][0],
-        )
-    except (KeyError, IndexError) as err:
-        raise InvalidResponse("Unable to parse login url") from err
-
-
-async def generate_token(client: ClientSession, refresh_token: str) -> token_data:
-    """Generate a token from the refresh token
-
-    :param client: async client for making request
-    :param refresh_token: Refresh token for generating request tokens
-    """
-    logger.debug("Generating token")
-    data = {
-        "grant_type": "refresh_token",
-        "refresh_token": refresh_token,
-        "scope": "openid email offline_access profile",
-        "client_id": HUBSPACE_DEFAULT_CLIENT_ID,
-    }
-    logger.debug(
-        ("URL: %s" "\n\tdata: %s" "\n\theaders: %s"),
-        HUBSPACE_TOKEN_URL,
-        data,
-        HUBSPACE_TOKEN_HEADERS,
-    )
-    async with client.post(
-        HUBSPACE_TOKEN_URL, headers=HUBSPACE_TOKEN_HEADERS, data=data
-    ) as response:
-        if response.status != 200:
-            with contextlib.suppress(ValueError, ContentTypeError):
-                data = await response.json()
-            if data and data.get("error") == "invalid_grant":
-                raise InvalidAuth()
-            else:
-                response.raise_for_status()
-        resp_json = await response.json()
-        try:
-            auth_token = resp_json["id_token"]
-        except KeyError:
-            raise InvalidResponse("Unable to extract the token")
-        logger.debug("JSON response: %s", resp_json)
-        return token_data(
-            auth_token, datetime.datetime.now().timestamp() + TOKEN_TIMEOUT
-        )
+__all__ = ["HubspaceAuth"]
+
+import asyncio
+import base64
+import contextlib
+import datetime
+import hashlib
+import logging
+import os
+import re
+from collections import namedtuple
+from typing import Final, Optional
+from urllib.parse import parse_qs, urlparse
+
+import aiohttp
+from aiohttp import ClientSession, ContentTypeError
+from bs4 import BeautifulSoup
+
+from ..errors import InvalidAuth, InvalidResponse
+from . import v1_const
+
+logger = logging.getLogger(__name__)
+
+TOKEN_TIMEOUT: Final[int] = 118
+STATUS_CODE: Final[str] = "Status Code: %s"
+
+auth_challenge = namedtuple("AuthChallenge", ["challenge", "verifier"])
+token_data = namedtuple("TokenData", ["token", "expiration"])
+auth_sess_data = namedtuple("AuthSessionData", ["session_code", "execution", "tab_id"])
+
+
+class HubspaceAuth:
+    """Authentication against the Hubspace API
+
+    This class follows the Hubspace authentication workflow and utilizes
+    refresh tokens.
+    """
+
+    def __init__(
```

Comment:
This class should probably be renamed to AferoAuth

Reply:
When I was writing the code, I wondered if I should rename all occurrences of "Hubspace".
It's a huge change... This PR is a start, to ask you if I'm heading in the right direction.

Reply:
I believe you are heading in the correct direction. I have been thinking about this and I think it makes a lot of sense to fork this into `aioafero` and we can remove the hubspace-specific things. Let me get started on that.

Reply:
Want to take a look at this? Tests pass so I am confident but wanted a second pair of eyes on it

https://github.com/Expl0dingBanana/aioafero/pull/1

Reply:
I see that you forked the homeassistant plugin too. is your plan to add myko to HA too? I am sure we could standardize that too

Reply:
Yes, that's my opinion too.